Improve torch::jit::as_{module,object} performance

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #72303
* #76420
* __->__ #84399

This caches the import of `torch.jit.ScriptModule`,
`torch.ScriptObject` and `torch.jit.RecursiveScriptClass`. I measure
a ~0.8 us performance uplift locally when calling a `torch.ops`
function with a `ScriptObject` argument.

peterbell10 committed:
Improve torch::jit::as_{module,object} performance

This caches the import of `torch.jit.ScriptModule` and
`torch.jit.RecursiveScriptObject`. I measure a ~0.8 us performance
uplift locally when calling a `torch.ops` function with a
`ScriptObject` argument.

[ghstack-poisoned]

ezyang reviewed (approved):
CI seems red, maybe?

peterbell10 committed:
Rebase on "Improve torch::jit::as_{module,object} performance"


This caches the import of `torch.jit.ScriptModule`,
`torch.ScriptObject` and `torch.jit.RecursiveScriptClass`. I measure
a ~0.8 us performance uplift locally when calling a `torch.ops`
function with a `ScriptObject` argument.


[ghstack-poisoned]

peterbell10 commented:
The CI failure looks unrelated to me:
```
OMP: Error #15: Initializing libiomp5.so, but found unknown library already initialized.
```

We'll see if it repeats.

peterbell10 committed:
Rebase on "Improve torch::jit::as_{module,object} performance"


This caches the import of `torch.jit.ScriptModule`,
`torch.ScriptObject` and `torch.jit.RecursiveScriptClass`. I measure
a ~0.8 us performance uplift locally when calling a `torch.ops`
function with a `ScriptObject` argument.


[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
